src/Modal/index.jsx

```diff
@@ -0,0 +1,167 @@
+    this.removeEvents = this.removeEvents.bind(this);
+    this.handleDocumentClick = this.handleDocumentClick.bind(this);
+    this.close = this.close.bind(this);
+    this.handleKeyDown = this.handleKeyDown.bind(this);
+    this.buttons = [];
+
+    // is there a way to avoid depending on the props for the open state?
+    this.state = {
+      open: props.open,
+      focusIndex: 0,
+    };
+  }
+
+  componentDidMount() {
+    this.buttons[0].focus();
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.open !== this.props.open) {
+      this.setState({ open: nextProps.open });
+    }
+  }
+
+  componentWillUpdate(_, nextState) {
+    if (nextState.open) {
+      this.addEvents();
+    } else {
+      this.removeEvents();
+    }
+  }
+
+  componentDidUpdate() {
+    if (this.state.open) {
+      this.buttons[this.state.focusIndex].focus();
+    }
+  }
+
+  close() {
+    this.setState({ open: false });
+    this.props.onClose();
+  }
+  // what does true mean here?
+  addEvents() {
+    document.addEventListener('click', this.handleDocumentClick, true);
+  }
+
+  removeEvents() {
+    document.removeEventListener('click', this.handleDocumentClick, true);
+  }
+
+  handleDocumentClick(e) {
+    const index = this.buttons.indexOf(e.target);
+    if (index > -1) {
+      this.setState({ focusIndex: index });
+    }
+  }
+
+  // open() {
+  //   this.setState({ open: true });
+  //   this.closeButton.focus();
+  // }
+
+  handleKeyDown(e) {
+    if (e.key === 'Escape') {
+      this.close();
+    } else if (e.key === 'Tab') {
+      e.preventDefault();
+      if (e.shiftKey) {
+        this.setState({
+          focusIndex: ((this.state.focusIndex - 1) + this.buttons.length) % this.buttons.length,
+        });
+      } else {
+        this.setState({
+          focusIndex: (this.state.focusIndex + 1) % this.buttons.length,
+        });
+      }
+    }
+  }
+
+  renderButtons() {
+    return this.props.buttons.map((button, i) => {
+      let buttonElement = button;
+      let buttonProps = button.props;
+
+      if (button.type !== Button) {
+        buttonProps = button;
+      }
+
+      buttonElement = (<Button
+        {...buttonProps}
+        key={i}
+        inputRef={(input) => { this.buttons[i + 1] = input; }}
+        onKeyDown={this.handleKeyDown}
+      />);
+
+      return buttonElement;
+    });
+  }
+
+  render() {
+    return (
+      <div className={classNames(styles.modal, styles.show, { [styles['modal-open']]: this.state.open })}>
+        <div className={styles['modal-dialog']} role="dialog" >
+          <div className={styles['modal-content']}>
+            <div className={styles['modal-header']}>
+              <h5 className={styles['modal-title']}> {this.props.title} </h5>
+              <Button
+                display="&times;"
+                aria-label="Close"
+                buttonType="light"
+                onClick={this.close}
+                inputRef={(input) => { this.buttons[0] = input; }}
```

Comment:
I assume the X button will always be the first button to be in focus, so I hard-coded the index here.